src/helpers/getFilteredExpenses.js

```diff
@@ -0,0 +1,58 @@
+'use strict';
+
+const getFilteredExpenses = (expenses, filter) => {
+  let filteredExpenses = [...expenses];
+
+  const {
+    userId,
+    categories,
+    from,
+    to,
+  } = filter;
+
+  let formattedCategories = [];
+
+  if (Array.isArray(filter.categories)) {
+    formattedCategories = [...categories];
+  } else if (categories) {
+    formattedCategories.push(categories);
+  }
+
+  if (userId) {
+    filteredExpenses = filteredExpenses
+      .filter(expense => expense.userId === +userId);
+  }
+
+  if (formattedCategories.length > 0) {
+    filteredExpenses = filteredExpenses
+      .filter(expense => categories.includes(expense.category));
+  }
+
+  if (from) {
+    const formattedFrom = new Date(from).getTime();
+
+    filteredExpenses = filteredExpenses
+      .filter(expense => {
+        const expenseTime = new Date(expense.spentAt).getTime();
+
+        return expenseTime >= formattedFrom;
+      });
+  }
+
+  if (to) {
+    const formattedTo = new Date(to).getTime();
+
+    filteredExpenses = filteredExpenses
+      .filter(expense => {
+        const expenseTime = new Date(expense.spentAt).getTime();
+
+        return expenseTime <= formattedTo;
+      });
+  }
+
+  return filteredExpenses;
+};
```

Comment:
Sorry, I haven't mentioned this the first time. This logic looks too complicated. We don't need to rerun filtering for each parameter. We can use all conditions inside one filter method
`expenses.filter(expense => {
    if (userId && expense.userId !== +userId) {
      return false;
    }
    
   // and so on 
    `